File: backend/internal/reporting/reports.go
// Package reporting delivers named financial reports differentiated by org kind.
//
// Business org reports:
//   - profit-and-loss  — income vs expense accounts over a period
//   - balance-sheet    — assets = liabilities + equity at a point in time
//   - vat-summary      — output vs input VAT via tax_rates
//
// Personal org reports:
//   - cash-flow        — cash in vs out aggregated by calendar month
//   - spending-trend   — spend by category over time
//   - net-worth        — asset/liability time series
//
// All report functions are pure: they take pre-fetched rows and return
// structured results with no I/O side effects.  Handlers fetch rows from
// the store then call these functions, making unit testing straightforward.
package reporting

import (
	"errors"
	"fmt"
	"time"
)

// ─── Errors ────────────────────────────────────────────────────────────────

// ErrWrongOrgKind is returned when a report is requested for an org kind
// that does not support it.
var ErrWrongOrgKind = errors.New("report not available for this org kind")

// ErrUnknownReport is returned when the report name is not recognised.
var ErrUnknownReport = errors.New("unknown report name")

// ─── Report registry ───────────────────────────────────────────────────────

// reportMeta describes a named report and the org kinds it supports.
type reportMeta struct {
	Name        string
	Description string
	Kinds       map[string]bool // "personal" | "business"
}

var registry = []reportMeta{
	{Name: "profit-and-loss", Description: "Income vs expense accounts over the period", Kinds: map[string]bool{"business": true}},
	{Name: "balance-sheet", Description: "Assets = liabilities + equity at the end date", Kinds: map[string]bool{"business": true}},
	{Name: "vat-summary", Description: "Output vs input VAT for the period", Kinds: map[string]bool{"business": true}},
	{Name: "cash-flow", Description: "Cash in vs out aggregated by month", Kinds: map[string]bool{"personal": true}},
	{Name: "spending-trend", Description: "Spend by category over time", Kinds: map[string]bool{"personal": true}},
	{Name: "net-worth", Description: "Net worth time series", Kinds: map[string]bool{"personal": true}},
}

// ValidateReport returns ErrUnknownReport if name is not in the registry, or
// ErrWrongOrgKind if orgKind does not support the requested report.
func ValidateReport(name, orgKind string) error {
	for _, m := range registry {
		if m.Name == name {
			if !m.Kinds[orgKind] {
				return fmt.Errorf("%w: %q is not available for %q orgs", ErrWrongOrgKind, name, orgKind)
			}
			return nil
		}
	}
	return fmt.Errorf("%w: %q", ErrUnknownReport, name)
}

// ─── Shared types ──────────────────────────────────────────────────────────

// Period is an inclusive date range.
type Period struct {
	From time.Time
	To   time.Time
}

// AccountLine is one row in a P&L or balance-sheet section.
type AccountLine struct {
	AccountID   string  `json:"account_id"`
	Code        string  `json:"code,omitempty"`
	Name        string  `json:"name"`
	AccountType string  `json:"account_type"`
	Balance     float64 `json:"balance"`
}

// ─── Profit & Loss ─────────────────────────────────────────────────────────

// PLInput is the set of ledger-derived rows needed to build a P&L.
// Each row has account type (income | expense) and a net balance for the
// period (credit-debit for income, debit-credit for expense).
type PLInput struct {
	Lines []PLLine
}

// PLLine is one account's contribution to the P&L.
type PLLine struct {
	AccountID   string
	Code        string
	Name        string
	AccountType string  // "income" | "expense"
	NetBalance  float64 // positive = normal-sign contribution
}

// PLReport is the result of BuildPL.
type PLReport struct {
	Period       Period        `json:"period"`
	IncomeLines  []AccountLine `json:"income_lines"`
	ExpenseLines []AccountLine `json:"expense_lines"`
	TotalIncome  float64       `json:"total_income"`
	TotalExpense float64       `json:"total_expense"`
	NetIncome    float64       `json:"net_income"`
}

// BuildPL produces a profit-and-loss statement from the provided ledger rows.
// Pure function — no I/O.
func BuildPL(p Period, rows []PLLine) PLReport {
	r := PLReport{Period: p}
	for _, l := range rows {
		line := AccountLine{
			AccountID:   l.AccountID,
			Code:        l.Code,
			Name:        l.Name,
			AccountType: l.AccountType,
			Balance:     l.NetBalance,
		}
		switch l.AccountType {
		case "income":
			r.IncomeLines = append(r.IncomeLines, line)
			r.TotalIncome += l.NetBalance
		case "expense":
			r.ExpenseLines = append(r.ExpenseLines, line)
			r.TotalExpense += l.NetBalance
		}
	}
	r.NetIncome = r.TotalIncome - r.TotalExpense
	return r
}

// ─── Balance Sheet ─────────────────────────────────────────────────────────

// BSLine is one account's contribution to the balance sheet.
type BSLine struct {
	AccountID   string
	Code        string
	Name        string
	AccountType string  // "asset" | "liability" | "equity"
	Balance     float64 // debit-positive for assets, credit-positive for L+E
}

// BSReport is the result of BuildBalanceSheet.
type BSReport struct {
	AsOf             time.Time     `json:"as_of"`
	AssetLines       []AccountLine `json:"asset_lines"`
	LiabilityLines   []AccountLine `json:"liability_lines"`
	EquityLines      []AccountLine `json:"equity_lines"`
	TotalAssets      float64       `json:"total_assets"`
	TotalLiabilities float64       `json:"total_liabilities"`
	TotalEquity      float64       `json:"total_equity"`
	// Check: TotalAssets should equal TotalLiabilities + TotalEquity.
	Balanced bool    `json:"balanced"`
	Diff     float64 `json:"diff"` // 0 when balanced within rounding
}

// BuildBalanceSheet assembles a balance-sheet report.
// For assets the normal sign is debit (positive Balance means debit).
// For liabilities and equity the normal sign is credit (positive Balance
// means credit).  Callers must pre-compute the signed balance per account.
// Pure function — no I/O.
func BuildBalanceSheet(asOf time.Time, rows []BSLine) BSReport {
	r := BSReport{AsOf: asOf}
	for _, l := range rows {
		line := AccountLine{
			AccountID:   l.AccountID,
			Code:        l.Code,
			Name:        l.Name,
			AccountType: l.AccountType,
			Balance:     l.Balance,
		}
		switch l.AccountType {
		case "asset":
			r.AssetLines = append(r.AssetLines, line)
			r.TotalAssets += l.Balance
		case "liability":
			r.LiabilityLines = append(r.LiabilityLines, line)
			r.TotalLiabilities += l.Balance
		case "equity":
			r.EquityLines = append(r.EquityLines, line)
			r.TotalEquity += l.Balance
		}
	}
	r.Diff = r.TotalAssets - (r.TotalLiabilities + r.TotalEquity)
	// Treat as balanced if diff is within 1 cent rounding.
	r.Balanced = r.Diff >= -0.01 && r.Diff <= 0.01
	return r
}

// ─── VAT Summary ────────────────────────────────────────────────────────────

// VATLine is one tax-rate's contribution to the VAT summary.
type VATLine struct {
	TaxRateID   string
	Code        string
	Name        string
	Rate        float64
	Net         float64 // taxable amount (excl tax)
	TaxAmount   float64 // collected / paid
	Direction   string  // "output" (sales) | "input" (purchases)
}

// VATReport is the result of BuildVAT.
type VATReport struct {
	Period        Period    `json:"period"`
	OutputLines   []VATLine `json:"output_lines"`
	InputLines    []VATLine `json:"input_lines"`
	TotalOutput   float64   `json:"total_output_tax"`
	TotalInput    float64   `json:"total_input_tax"`
	NetVATPayable float64   `json:"net_vat_payable"`
}

// BuildVAT assembles a VAT summary from pre-fetched tax-rate lines.
// Pure function — no I/O.
func BuildVAT(p Period, rows []VATLine) VATReport {
	r := VATReport{Period: p}
	for _, l := range rows {
		switch l.Direction {
		case "output":
			r.OutputLines = append(r.OutputLines, l)
			r.TotalOutput += l.TaxAmount
		case "input":
			r.InputLines = append(r.InputLines, l)
			r.TotalInput += l.TaxAmount
		}
	}
	r.NetVATPayable = r.TotalOutput - r.TotalInput
	return r
}

// ─── Cash Flow ─────────────────────────────────────────────────────────────

// CashFlowMonth is one calendar month's in/out totals.
type CashFlowMonth struct {
	Month   string  `json:"month"` // "YYYY-MM"
	Inflow  float64 `json:"inflow"`
	Outflow float64 `json:"outflow"`
	Net     float64 `json:"net"`
}

// CashFlowInput is one transaction row used to build the cash flow report.
type CashFlowInput struct {
	Month     string  // "YYYY-MM"
	Direction string  // "credit" | "debit" | "transfer"
	Amount    float64
}

// CashFlowReport is the result of BuildCashFlow.
type CashFlowReport struct {
	Period      Period          `json:"period"`
	Months      []CashFlowMonth `json:"months"`
	TotalInflow float64         `json:"total_inflow"`
	TotalOutflow float64        `json:"total_outflow"`
	NetCashFlow float64         `json:"net_cash_flow"`
}

// BuildCashFlow produces a monthly cash-flow report.
// Credits = inflows (money coming in), debits = outflows.
// Transfers are ignored to avoid double-counting.
// Pure function — no I/O.
func BuildCashFlow(p Period, rows []CashFlowInput) CashFlowReport {
	type monthly struct {
		in  float64
		out float64
	}
	// Ordered month keys.
	order := make([]string, 0)
	byMonth := make(map[string]*monthly)

	for _, row := range rows {
		if row.Direction == "transfer" {
			continue
		}
		m, exists := byMonth[row.Month]
		if !exists {
			m = &monthly{}
			byMonth[row.Month] = m
			order = append(order, row.Month)
		}
		switch row.Direction {
		case "credit":
			m.in += row.Amount
		case "debit":
			m.out += row.Amount
		}
	}

	// Deduplicate order (in case rows arrive unsorted).
	seen := make(map[string]bool)
	uniqueOrder := make([]string, 0, len(order))
	for _, k := range order {
		if !seen[k] {
			seen[k] = true
			uniqueOrder = append(uniqueOrder, k)
		}
	}
	// Sort months lexicographically ("YYYY-MM" sorts correctly).
	sortStrings(uniqueOrder)

	r := CashFlowReport{Period: p}
	for _, mon := range uniqueOrder {
		m := byMonth[mon]
		net := m.in - m.out
		r.Months = append(r.Months, CashFlowMonth{
			Month:   mon,
			Inflow:  m.in,
			Outflow: m.out,
			Net:     net,
		})
		r.TotalInflow += m.in
		r.TotalOutflow += m.out
	}
	r.NetCashFlow = r.TotalInflow - r.TotalOutflow
	return r
}

// ─── Spending Trend ─────────────────────────────────────────────────────────

// SpendingTrendRow is one (category, month) cell.
type SpendingTrendRow struct {
	CategoryID   string  `json:"category_id"`
	CategoryName string  `json:"category_name"`
	Month        string  `json:"month"` // "YYYY-MM"
	Amount       float64 `json:"amount"`
}

// SpendingTrendInput feeds BuildSpendingTrend.
type SpendingTrendInput struct {
	CategoryID   string
	CategoryName string
	Month        string
	Amount       float64
}

// SpendingTrendReport is the result of BuildSpendingTrend.
type SpendingTrendReport struct {
	Period Period             `json:"period"`
	Rows   []SpendingTrendRow `json:"rows"`
	// Months lists the unique months across all rows (ordered).
	Months []string `json:"months"`
}

// BuildSpendingTrend assembles the spending-trend matrix.
// Pure function — no I/O.
func BuildSpendingTrend(p Period, rows []SpendingTrendInput) SpendingTrendReport {
	monthSet := map[string]bool{}
	out := make([]SpendingTrendRow, 0, len(rows))
	for _, row := range rows {
		out = append(out, SpendingTrendRow{
			CategoryID:   row.CategoryID,
			CategoryName: row.CategoryName,
			Month:        row.Month,
			Amount:       row.Amount,
		})
		monthSet[row.Month] = true
	}
	months := make([]string, 0, len(monthSet))
	for m := range monthSet {
		months = append(months, m)
	}
	sortStrings(months)
	return SpendingTrendReport{Period: p, Rows: out, Months: months}
}

// ─── Net Worth ──────────────────────────────────────────────────────────────

// NetWorthPoint is the net worth at a single point in time.
type NetWorthPoint struct {
	Date        string  `json:"date"` // "YYYY-MM-DD"
	TotalAssets float64 `json:"total_assets"`
	TotalDebt   float64 `json:"total_debt"`
	NetWorth    float64 `json:"net_worth"`
}

// NetWorthInput feeds BuildNetWorth.
type NetWorthInput struct {
	Date        string
	TotalAssets float64
	TotalDebt   float64
}

// NetWorthReport is the result of BuildNetWorth.
type NetWorthReport struct {
	Period Period          `json:"period"`
	Series []NetWorthPoint `json:"series"`
}

// BuildNetWorth assembles the net-worth time series.
// Pure function — no I/O.
func BuildNetWorth(p Period, rows []NetWorthInput) NetWorthReport {
	series := make([]NetWorthPoint, 0, len(rows))
	for _, r := range rows {
		series = append(series, NetWorthPoint{
			Date:        r.Date,
			TotalAssets: r.TotalAssets,
			TotalDebt:   r.TotalDebt,
			NetWorth:    r.TotalAssets - r.TotalDebt,
		})
	}
	return NetWorthReport{Period: p, Series: series}
}

// ─── Utility ───────────────────────────────────────────────────────────────

// sortStrings sorts a slice of strings in ascending order (stdlib sort-free
// insertion sort — avoids importing "sort" for a tiny slice).
func sortStrings(ss []string) {
	for i := 1; i < len(ss); i++ {
		key := ss[i]
		j := i - 1
		for j >= 0 && ss[j] > key {
			ss[j+1] = ss[j]
			j--
		}
		ss[j+1] = key
	}
}

File: worker/src/modules/reporting/reports.ts
/**
 * Financial report builders — pure functions, no I/O.
 * Port of Go internal/reporting/reports.go.
 *
 * MONEY INVARIANT: all arithmetic goes through lib/money (Decimal). Balances
 * are returned as money() strings (2 dp). Never use JS number for currency.
 */
import { dec, money, sum } from "../../lib/money";
import type {
  Period,
  ReportMeta,
  AccountLine,
  PLLine,
  PLReport,
  BSLine,
  BSReport,
  VATLine,
  VATReport,
  CashFlowRow,
  CashFlowMonth,
  CashFlowReport,
  SpendingTrendRow,
  SpendingTrendReport,
  NetWorthInput,
  NetWorthPoint,
  NetWorthReport,
} from "./types";
import { REGISTRY } from "./types";

// ─── Errors ──────────────────────────────────────────────────────────────────

export class ErrUnknownReport extends Error {
  constructor(name: string) {
    super(`unknown report name: "${name}"`);
    this.name = "ErrUnknownReport";
  }
}

export class ErrWrongOrgKind extends Error {
  constructor(report: string, kind: string) {
    super(`report not available for this org kind: "${report}" is not available for "${kind}" orgs`);
    this.name = "ErrWrongOrgKind";
  }
}

/**
 * Validate that `name` exists and is available for `orgKind`.
 * Throws ErrUnknownReport or ErrWrongOrgKind — callers map to HTTP status.
 */
export function validateReport(name: string, orgKind: string): ReportMeta {
  const meta = REGISTRY.find((m) => m.name === name);
  if (!meta) throw new ErrUnknownReport(name);
  if (!meta.kinds.has(orgKind)) throw new ErrWrongOrgKind(name, orgKind);
  return meta;
}

// ─── Profit & Loss ─────────────────────────────────────────────────────────

/**
 * BuildPL — port of Go BuildPL.
 * income: credit – debit (normal credit-side balance, positive).
 * expense: debit – credit (normal debit-side balance, positive).
 * netIncome = totalIncome − totalExpense (negative = loss).
 */
export function buildPL(period: Period, rows: PLLine[]): PLReport {
  const incomeLines: AccountLine[] = [];
  const expenseLines: AccountLine[] = [];
  let totalIncome = dec(0);
  let totalExpense = dec(0);

  for (const l of rows) {
    const line: AccountLine = {
      account_id: l.account_id,
      code: l.code,
      name: l.name,
      account_type: l.account_type,
      balance: money(l.net_balance),
    };
    if (l.account_type === "income") {
      incomeLines.push(line);
      totalIncome = totalIncome.plus(dec(l.net_balance));
    } else if (l.account_type === "expense") {
      expenseLines.push(line);
      totalExpense = totalExpense.plus(dec(l.net_balance));
    }
  }

  const netIncome = totalIncome.minus(totalExpense);

  return {
    period,
    income_lines: incomeLines,
    expense_lines: expenseLines,
    total_income: money(totalIncome),
    total_expense: money(totalExpense),
    net_income: money(netIncome),
  };
}

// ─── Balance Sheet ──────────────────────────────────────────────────────────

/**
 * BuildBalanceSheet — port of Go BuildBalanceSheet.
 * Assets: debit – credit (positive = asset value).
 * Liabilities/Equity: credit – debit (positive = owed/invested).
 * balanced = |assets − (liabilities + equity)| ≤ 0.01.
 */
export function buildBalanceSheet(asOf: string, rows: BSLine[]): BSReport {
  const assetLines: AccountLine[] = [];
  const liabilityLines: AccountLine[] = [];
  const equityLines: AccountLine[] = [];
  let totalAssets = dec(0);
  let totalLiabilities = dec(0);
  let totalEquity = dec(0);

  for (const l of rows) {
    const line: AccountLine = {
      account_id: l.account_id,
      code: l.code,
      name: l.name,
      account_type: l.account_type,
      balance: money(l.balance),
    };
    switch (l.account_type) {
      case "asset":
        assetLines.push(line);
        totalAssets = totalAssets.plus(dec(l.balance));
        break;
      case "liability":
        liabilityLines.push(line);
        totalLiabilities = totalLiabilities.plus(dec(l.balance));
        break;
      case "equity":
        equityLines.push(line);
        totalEquity = totalEquity.plus(dec(l.balance));
        break;
    }
  }

  const diff = totalAssets.minus(totalLiabilities.plus(totalEquity));
  const balanced = diff.abs().lte(dec("0.01"));

  return {
    as_of: asOf,
    asset_lines: assetLines,
    liability_lines: liabilityLines,
    equity_lines: equityLines,
    total_assets: money(totalAssets),
    total_liabilities: money(totalLiabilities),
    total_equity: money(totalEquity),
    balanced,
    diff: money(diff),
  };
}

// ─── VAT Summary ─────────────────────────────────────────────────────────────

/**
 * BuildVAT — port of Go BuildVAT.
 * output = tax collected on sales; input = tax paid on purchases.
 * netVATPayable = totalOutput − totalInput (negative = refund due).
 */
export function buildVAT(period: Period, rows: VATLine[]): VATReport {
  const outputLines: VATLine[] = [];
  const inputLines: VATLine[] = [];
  let totalOutput = dec(0);
  let totalInput = dec(0);

  for (const l of rows) {
    if (l.direction === "output") {
      outputLines.push(l);
      totalOutput = totalOutput.plus(dec(l.tax_amount));
    } else if (l.direction === "input") {
      inputLines.push(l);
      totalInput = totalInput.plus(dec(l.tax_amount));
    }
  }

  return {
    period,
    output_lines: outputLines,
    input_lines: inputLines,
    total_output_tax: money(totalOutput),
    total_input_tax: money(totalInput),
    net_vat_payable: money(totalOutput.minus(totalInput)),
  };
}

// ─── Cash Flow ───────────────────────────────────────────────────────────────

/**
 * BuildCashFlow — port of Go BuildCashFlow.
 * Credits = inflows; debits = outflows. Transfers ignored (avoid double-count).
 * Months sorted lexicographically (YYYY-MM sorts correctly).
 */
export function buildCashFlow(period: Period, rows: CashFlowRow[]): CashFlowReport {
  const byMonth = new Map<string, { in: ReturnType<typeof dec>; out: ReturnType<typeof dec> }>();
  const order: string[] = [];

  for (const row of rows) {
    if (row.direction === "transfer") continue;
    if (!byMonth.has(row.month)) {
      byMonth.set(row.month, { in: dec(0), out: dec(0) });
      order.push(row.month);
    }
    const m = byMonth.get(row.month)!;
    if (row.direction === "credit") {
      m.in = m.in.plus(dec(row.amount));
    } else if (row.direction === "debit") {
      m.out = m.out.plus(dec(row.amount));
    }
  }

  // Deduplicate and sort months.
  const uniqueMonths = [...new Set(order)].sort();

  const months: CashFlowMonth[] = [];
  let totalInflow = dec(0);
  let totalOutflow = dec(0);

  for (const mon of uniqueMonths) {
    const m = byMonth.get(mon)!;
    const net = m.in.minus(m.out);
    months.push({
      month: mon,
      inflow: money(m.in),
      outflow: money(m.out),
      net: money(net),
    });
    totalInflow = totalInflow.plus(m.in);
    totalOutflow = totalOutflow.plus(m.out);
  }

  return {
    period,
    months,
    total_inflow: money(totalInflow),
    total_outflow: money(totalOutflow),
    net_cash_flow: money(totalInflow.minus(totalOutflow)),
  };
}

// ─── Spending Trend ──────────────────────────────────────────────────────────

/**
 * BuildSpendingTrend — port of Go BuildSpendingTrend.
 * Returns per-(category, month) rows plus the sorted unique month list.
 */
export function buildSpendingTrend(period: Period, rows: SpendingTrendRow[]): SpendingTrendReport {
  const monthSet = new Set<string>();
  const out: SpendingTrendRow[] = [];

  for (const row of rows) {
    out.push({
      category_id: row.category_id,
      category_name: row.category_name,
      month: row.month,
      amount: money(row.amount),
    });
    monthSet.add(row.month);
  }

  const months = [...monthSet].sort();

  return { period, rows: out, months };
}

// ─── Net Worth ───────────────────────────────────────────────────────────────

/**
 * BuildNetWorth — port of Go BuildNetWorth.
 * net_worth = total_assets − total_debt per data point.
 */
export function buildNetWorth(period: Period, rows: NetWorthInput[]): NetWorthReport {
  const series: NetWorthPoint[] = rows.map((r) => ({
    date: r.date,
    total_assets: money(r.total_assets),
    total_debt: money(r.total_debt),
    net_worth: money(dec(r.total_assets).minus(dec(r.total_debt))),
  }));

  return { period, series };
}

// Re-export sum for use in tests (tree-shaken away if unused).
export { sum };
